semver/src/lib.sk

```diff
@@ -0,0 +1,267 @@
+    )
+  }
+}
+
+class VersionReq(comparators: Array<Comparator>) uses Show {
+  const kStar: VersionReq = VersionReq(Array[]);
+
+  fun matches(version: Version): Bool {
+    for (cmp in this.comparators) {
+      if (!cmp.matches(version)) {
+        return false
+      }
+    };
+
+    if (version.pre.isEmpty()) {
+      return true
+    };
+
+    for (cmp in this.comparators) {
+      if (pre_is_compatible(cmp, version)) {
+        return true
+      }
+    };
+
+    false
+  }
+
+  static fun fromString(str: String): VersionReq {
+    !str = str.trim();
+    if (str.startsWith("*") || str.startsWith("x") || str.startsWith("X")) {
+      if (str != "*" && str != "x" && str != "X") {
+        invariant_violation("Wildcard must be the only operator")
+      };
+      return static::kStar
+    };
+
+    comparators = str
+      .split(",")
+      .map(cmp -> Comparator::fromString(cmp))
+      .collect(Array);
+
+    VersionReq(comparators)
+  }
+
+  fun toString(): String {
+    if (this.comparators.isEmpty()) {
+      return "*"
+    };
+    this.comparators.map(cmp -> cmp.toString()).join(", ")
+  }
+}
+
+class Comparator(
+  op: Op,
+  major: Int,
+  minor: ?Int,
+  patch: ?Int,
+  pre: Prerelease,
+) uses Show {
+  fun matches(version: Version): Bool {
+    this.op match {
+    | Exact()
+    | Wildcard() ->
+      matches_exact(this, version)
+    | Greater() -> matches_greater(this, version)
+    | GreaterEq() ->
+      matches_exact(this, version) || matches_greater(this, version)
+    | Less() -> matches_less(this, version)
+    | LessEq() -> matches_exact(this, version) || matches_less(this, version)
+    | Tilde() -> matches_tilde(this, version)
+    | Caret() -> matches_caret(this, version)
+    }
+  }
+
+  static fun fromString(str: String): Comparator {
+    pos = Lexer.LexingPosition::create(str.trim());
+
+    res = parse_comparator(pos);
+
+    if (!pos.atEnd()) {
+      invariant_violation("Unexpected characters after comparator")
+    };
+
+    res
+  }
+
+  fun toString(): String {
+    prefix = this.op match {
+    | Exact() -> "="
+    | Greater() -> ">"
+    | GreaterEq() -> ">="
+    | Less() -> "<"
+    | LessEq() -> "<="
+    | Tilde() -> "~"
+    | Caret() -> "^"
+    | Wildcard() -> ""
+    };
+    `${prefix}${this.major}` +
+      this.minor match {
+      | Some(minor) ->
+        `.${minor}` +
+          this.patch match {
+          | Some(patch) ->
+            `.${patch}` +
+              if (!this.pre.isEmpty()) {
+                `-${this.pre}`
+              } else {
+                ""
+              }
+          | None() if (this.op is Wildcard _) -> ".*"
+          | None() -> ""
+          }
+      | None() if (this.op is Wildcard _) -> ".*"
+      | _ -> ""
+      }
+  }
+}
+
+base class Op {
+  children =
+  | Exact()
+  | Greater()
+  | GreaterEq()
+  | Less()
+  | LessEq()
+  | Tilde()
+  | Caret()
+  | Wildcard()
+}
+
+class Prerelease(identifier: String) uses Show, Orderable {
```

Comment:
That's an option, although there is a tradeoff between processing the pre-release tag upfront (possibly for a large number of elements, for instance when parsing a package repository index), and doing it dynamically (and each time, though it could be cached) during comparison. Since in the vast majority of package version comparisons the pre-release tag won't even come into play, I opted to not pay the cost upfront (not that perfs matter all that much here at the moment).